bfsaz/bfsaz.go

```diff
@@ -0,0 +1,414 @@
+		Metadata:    bfs.NormMetadata(transKeys(resp.NewMetadata(), "_", "-")),
+	}, nil
+}
+
+// Open implements bfs.Bucket.
+func (b *bucket) Open(ctx context.Context, name string) (bfs.Reader, error) {
+	resp, err := b.NewBlockBlobURL(b.withPrefix(name)).
+		Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &reader{
+		ReadCloser:    resp.Body(azblob.RetryReaderOptions{}),
+		ContentLength: resp.ContentLength(),
+	}, nil
+}
+
+// Create implements bfs.Bucket.
+func (b *bucket) Create(ctx context.Context, name string, opts *bfs.WriteOptions) (bfs.Writer, error) {
+	f, err := ioutil.TempFile("", "bfs-az")
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		File: f,
+		ctx:  ctx,
+		blob: b.NewBlockBlobURL(b.withPrefix(name)),
+		opts: opts,
+	}, nil
+}
+
+// Remove implements bfs.Bucket.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	_, err := b.NewBlockBlobURL(b.withPrefix(name)).
+		Delete(ctx, "", azblob.BlobAccessConditions{})
+	if ne := normError(err); ne != nil && ne != bfs.ErrNotFound {
+		return ne
+	}
+	return nil
+}
+
+// Close implements bfs.Bucket.
+func (*bucket) Close() error { return nil }
+
+// --------------------------------------------------------------------
+
+func normError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var se azblob.StorageError
+	if errors.As(err, &se) {
+		if se.ServiceCode() == azblob.ServiceCodeBlobNotFound {
+			return bfs.ErrNotFound
+		}
+		return err
+	}
+
+	if pe := pipeline.Cause(err); pe != nil {
+		if ue := errors.Unwrap(pe); ue != nil {
+			return ue
+		}
+		return pe
+	}
+
+	return err
+}
+
+func transKeys(meta map[string]string, from, to string) map[string]string {
+	for k, v := range meta {
+		if nk := strings.ReplaceAll(k, from, to); nk != k {
+			delete(meta, k)
+			meta[nk] = v
+		}
+	}
+	return meta
+}
+
+// --------------------------------------------------------------------
+
+type writer struct {
+	*os.File
+
+	ctx  context.Context
+	blob azblob.BlockBlobURL
+	opts *bfs.WriteOptions
+
+	closeOnce sync.Once
+}
+
+func (w *writer) Discard() error {
+	err := context.Canceled
+	w.closeOnce.Do(func() {
+		// Delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// Close tempfile
+		err = w.File.Close()
+	})
+
+	return err
+}
+
+func (w *writer) Commit() error {
+	err := context.Canceled
+	w.closeOnce.Do(func() {
+		// Delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// Close tempfile
+		if err = w.File.Close(); err != nil {
+			return
+		}
+
+		// Re-open tempfile for reading
+		var file *os.File
+		if file, err = os.Open(fname); err != nil {
+			return
+		}
+		defer file.Close()
+
+		// Upload file
+		_, err = w.blob.Upload(
+			w.ctx, file,
+			azblob.BlobHTTPHeaders{
+				ContentType: w.opts.GetContentType(),
+			},
+			azblob.Metadata(transKeys(w.opts.GetMetadata(), "-", "_")),
```

Comment:
@mxmCherry same again: `GetMetadata()` copies the map and canonicalises the keys, then transKeys is applied on top